File: experiments/control-run/go/jobq/contract/contract.go
// Package contract is Mo's requires, ensures, and invariant as plain checks.
// A failed check returns an error that carries the condition's text; it
// never panics, so a contract on input is an error the caller handles.
package contract

// Violation is a failed contract: which kind, and the condition's text.
type Violation struct {
	Kind string
	Cond string
}

func (v *Violation) Error() string { return v.Kind + " failed: " + v.Cond }

// Require checks a precondition on a function's input.
func Require(ok bool, cond string) error { return check("requires", ok, cond) }

// Ensure checks a postcondition on a function's result.
func Ensure(ok bool, cond string) error { return check("ensures", ok, cond) }

// Invariant checks a property that holds after every operation.
func Invariant(ok bool, cond string) error { return check("invariant", ok, cond) }

func check(kind string, ok bool, cond string) error {
	if ok {
		return nil
	}
	return &Violation{Kind: kind, Cond: cond}
}

File: experiments/control-run/python/jobq/src/jobq/contract.py
"""Contracts as explicit checks. A failure names the condition that did not hold.

These are not `assert` statements, so `python -O` does not remove them.
"""


class ContractError(Exception):
    """A requires, ensures, or invariant check failed: a bug in jobq, not bad input."""


def require(condition: bool, text: str) -> None:
    """A precondition the caller must meet."""
    if not condition:
        raise ContractError(f"requires {text}")


def ensure(condition: bool, text: str) -> None:
    """A postcondition the function promises."""
    if not condition:
        raise ContractError(f"ensures {text}")


def invariant(condition: bool, text: str) -> None:
    """A condition that holds after every change of state."""
    if not condition:
        raise ContractError(f"invariant {text}")


def never(condition: bool, text: str) -> None:
    """A condition no change of state may ever break, checked on every change."""
    if not condition:
        raise ContractError(f"never broken: {text}")


class BoardFailure(Exception):
    """The board failed in a way that is not one request's: a rule broken inside the queue,
    or an unexpected error while a change was being applied. The board is rebuilt from the
    log; the request that met the failure is a 503."""


class ChaosFailure(BoardFailure):
    """The failure `--crash-every` injects: after a write is on disk, before it is applied."""
